internal/httputil/seek.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package httputil
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// readSeekCloser seeks http body by starting new connections.
+type readSeekCloser struct {
+	client *http.Client
+	req    *http.Request
+	rc     io.ReadCloser
+	size   int64
+	offset int64
+	closed bool
+}
+
+// NewReadSeekCloser returns a seeker to make the HTTP response seekable.
+// Callers should ensure that the server supports Range request.
+func NewReadSeekCloser(client *http.Client, req *http.Request, respBody io.ReadCloser, size int64) io.ReadSeekCloser {
+	return &readSeekCloser{
+		client: client,
+		req:    req,
+		rc:     respBody,
+		size:   size,
+	}
+}
+
+// Read reads the content body and counts offset.
+func (rsc *readSeekCloser) Read(p []byte) (n int, err error) {
+	n, err = rsc.rc.Read(p)
+	rsc.offset += int64(n)
```

Comment:
Should this be done only when err != nil?

Reply:
According to the [io.Reader](https://pkg.go.dev/io#Reader) doc,

> Callers should always process the n > 0 bytes returned before considering the error err. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors.

Reply:
Here is the `Read()` implementation for `io.LimitedReader` from the go built-in library for reference purpose.

```golang
func (l *LimitedReader) Read(p []byte) (n int, err error) {
	if l.N <= 0 {
		return 0, EOF
	}
	if int64(len(p)) > l.N {
		p = p[0:l.N]
	}
	n, err = l.R.Read(p)
	l.N -= int64(n)
	return
}
```

Reply:
If `err == nil`, then `n` will be `len(p)`. We still need to add it to the `offset`.

Instead of
```golang
func (rsc *readSeekCloser) Read(p []byte) (n int, err error) {
	n, err = rsc.rc.Read(p)
	if err != nil {
		rsc.offset += int64(n)
		return
	}
	rsc.offset += int64(n)
	return
}
```
we can simplify as
```golang
func (rsc *readSeekCloser) Read(p []byte) (n int, err error) {
	n, err = rsc.rc.Read(p)
	rsc.offset += int64(n)
	return
}
```